src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,538 @@
+        public bool SessionMode
+        {
+            get
+            {
+                return this.sessionMode;
+            }
+            set
+            {
+                this.sessionMode = value;
+            }
+        }
+
+        public SecuritySessionServerSettings SessionServerSettings
+        {
+            get
+            {
+                if (this.sessionServerSettings == null)
+                {
+                    lock (ThisLock)
+                    {
+                        if (this.sessionServerSettings == null)
+                        {
+                            SecuritySessionServerSettings tmp = new SecuritySessionServerSettings();
+                            System.Threading.Thread.MemoryBarrier();
+                            tmp.SecurityServiceDispatcher = this;
+                            this.sessionServerSettings = tmp;
+                        }
+                    }
+                }
+                return this.sessionServerSettings;
+            }
+        }
+
+        private bool SupportsDuplex
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return this.securityProtocolFactory.SupportsDuplex;
+            }
+        }
+
+        private bool SupportsRequestReply
+        {
+            get
+            {
+                ThrowIfProtocolFactoryNotSet();
+                return this.securityProtocolFactory.SupportsRequestReply;
+            }
+        }
+
+        public bool SendUnsecuredFaults
+        {
+            get
+            {
+                return this.sendUnsecuredFaults;
+            }
+            set
+            {
+                this.sendUnsecuredFaults = value;
+            }
+        }
+
+        public IChannel OuterChannel
+        {
+            get { return this.outerChannel; }
+            set { this.outerChannel = value; }
+        }
+        public Type AcceptorChannelType
+        {
+            get { return acceptorChannelType; }
+            set { acceptorChannelType = value; }
+        }
+
+        IList<Type> IServiceDispatcher.SupportedChannelTypes => throw new NotImplementedException();
+
+        public ServiceHostBase Host =>  InnerServiceDispatcher.Host;
+
+        internal void InitializeSecurityDispatcher(ChannelBuilder channelBuilder, Type type)
+        {
+            this.channelBuilder = channelBuilder;
+
+            if (this.SessionMode)
+            {
+                this.sessionServerSettings.ChannelBuilder = this.ChannelBuilder;
+                //this.InnerChannelListener = this.sessionServerSettings.CreateInnerChannelListener();
+                // this.Acceptor = this.sessionServerSettings.CreateAcceptor<TChannel>();
+                this.acceptorChannelType = type;
+                this.sessionServerSettings.AcceptorChannelType = type;
+            }
+            else
+            {
+               // this.InnerChannelListener = this.ChannelBuilder.BuildChannelListener<TChannel>();
```

Comment:
What happens if SessionMode is false? If this isn't currently supported, we should throw PNSE rather than have requests go nowhere

Reply:
added PNSE